dev/conductor/core/lib/src/codesign.dart

```diff
@@ -320,96 +390,789 @@ class CodesignCommand extends Command<void> {
+
+  int _remoteDownloadIndex = 0;
+  int get remoteDownloadIndex => _remoteDownloadIndex++;
+
+  static const String _entitlementsFileContents = '''
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE plist PUBLIC "-//Apple Computer//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
+<plist version="1.0">
+    <dict>
+        <key>com.apple.security.cs.allow-jit</key>
+        <true/>
+        <key>com.apple.security.cs.allow-unsigned-executable-memory</key>
+        <true/>
+        <key>com.apple.security.cs.allow-dyld-environment-variables</key>
+        <true/>
+        <key>com.apple.security.network.client</key>
+        <true/>
+        <key>com.apple.security.network.server</key>
+        <true/>
+        <key>com.apple.security.cs.disable-library-validation</key>
+        <true/>
+    </dict>
+</plist>
+''';
+
+  /// A [Map] from SHA1 hash of file contents to file pathname from expected
+  /// files to codesign.
+  ///
+  /// These will be cross-referenced with all binary files unzipped.
+  final Map<String, String> expectedFileHashes = <String, String>{};
+
+  /// A [Map] from SHA1 hash of file contents to file pathname of codesigned
+  /// binary files.
+  ///
+  /// This will be used to cross reference the remote files listed in
+  /// [RemoteZip.archives] with the results of calling `flutter precache`.
+  final Map<String, String> codesignedFileHashes = <String, String>{};
+
+  /// A [Map] from SHA1 hash of file contents to file pathname of actual
+  /// downloaded binary files.
+  final Map<String, String> actualFileHashes = <String, String>{};
+
+  int _nextId = 0;
+  int get nextId {
+    final int currentKey = _nextId;
+    _nextId += 1;
+    return currentKey;
+  }
+
+  Future<void> validateAll(Iterable<RemoteZip> archives) async {
+    final Iterable<Future<void>> futures = archives.map<Future<void>>((RemoteZip archive) {
+      final Directory outDir = tempDir.childDirectory('remote_zip_$nextId');
+      return archive.visit(this, outDir);
+    });
+    await Future.wait(
+      futures,
+      eagerError: true,
     );
-    final List<String> allFiles = (result.stdout as String)
-        .split('\n')
-        .where((String s) => s.isNotEmpty)
-        .toList();
+    _validateFileHashes();
+    // TODO messaging?
+  }
 
-    await Future.forEach(allFiles, (String filePath) async {
-      if (await isBinary(filePath)) {
-        allBinaryPaths.add(filePath);
-      }
+  /// Unzip an [EmbeddedZip] and visit its children.
+  ///
+  /// The [parent] directory is scoped to the parent zip file.
+  @override
+  Future<void> visitEmbeddedZip(EmbeddedZip file, Directory parent) async {
+    final File localFile = await _validateFileExists(file, parent);
+    final Directory newDir = tempDir.childDirectory('embedded_zip_$nextId');
+    final package_arch.Archive? archive = await _unzip(localFile, newDir);
+    if (archive != null) {
+      await _hashActualFiles(archive, newDir);
```

Comment:
hashActualFiles seem to only be run when archive isn't null, which means the download has failed. So if the download succeeds, then the actual hashes will not be recorded

Reply:
Thanks!

Reply:
Christopher is too polite QAQ